apps/prairielearn/elements/pl-external-grader-results/pl-external-grader-results.py

```diff
@@ -39,129 +41,140 @@ def ansi_to_html(output):
+        html_params["achieved_max_points"] = results.get("score", 0) >= 1.0
+        html_params["results_color"] = (
+            PLColor("correct_green")
+            if (results.get("score", 0) >= 1.0)
+            else PLColor("incorrect_red")
+        )
+
+        html_params["message"] = ansi_to_html(results.get("message", None))
+        html_params["has_message"] = bool(html_params["message"])
+
+        html_params["output"] = ansi_to_html(results.get("output", None))
+        html_params["has_output"] = bool(html_params["output"])
+
+        html_params["images"] = results.get("images", [])
+        html_params["has_images"] = bool(html_params["images"])
+
+        html_params["has_message_or_output_or_image"] = (
+            html_params["has_message"]
+            or html_params["has_output"]
+            or html_params["has_images"]
+        )
+
+        results_tests = results.get("tests", None)
+        html_params["has_tests"] = bool(results_tests)
+
+        if results_tests:
+            # Let's not assume that people give us a valid array of tests
+            # If any test is missing either points or max_points, we'll
+            # disable detailed scores for all questions
+            tests_missing_points = any(
+                [
+                    test.get("points", None) is None
+                    or test.get("max_points", None) is None
+                    for test in results_tests
+                ]
+            )
+            html_params["tests_missing_points"] = tests_missing_points
+
+            if not tests_missing_points:
+                points_sum = sum(test["points"] for test in results_tests)
+                html_params["points"] = round_value(points_sum)
+
+                max_points_sum = sum(test["max_points"] for test in results_tests)
+                html_params["max_points"] = round_value(max_points_sum)
+
+            # We need to build a new tests array to massage data a bit
+            Test = dict[str, str | bool | int | PLColor | None]
+            tests: list[Test] = []
+            for index, results_test in enumerate(results_tests):
+                name = results_test.get("name", "")
+                message = results_test.get("message", None)
+                output = results_test.get("output", None)
+                description = results_test.get("description", None)
+
+                test: Test = {
+                    "index": index,
+                    "name": name,
+                    "message": message,
+                    "has_message": bool(message),
+                    "output": output,
+                    "has_output": bool(output),
+                    "description": description,
+                    "has_description": bool(description),
+                    "show_points": not tests_missing_points,
+                }
+
+                if not tests_missing_points:
+                    test["points"] = round_value(results_test.get("points"))
```

Comment:
Looking over the code, I'm not opposed to using string comparison to award points (as otherwise, we could wind up with weird discrepancies between the display and points awarded). Not sure if others have objections however.

Reply:
That's what were doing currently right? As mentioned in:

```plaintext
# Don't consider points for test cases that are 0/0
# to be correct. We compare with a string because
# `round_value()` returns a string.
```

Reply:
Ahh ok, yeah if this is the existing logic then this is fine.